core/src/main/c/share/txn_board.cpp

```diff
@@ -25,127 +25,127 @@
 #include <cstdint>
 #include "util.h"
 #include "jni.h"
+#include <atomic>
+#include <algorithm>
 
 #define COUNTER_T uint16_t
 
-template<typename T>
-void set_max_atomic(T *slot, T value) {
-    do {
-        T current = __atomic_load_n(slot, __ATOMIC_RELAXED);
-        if (value <= current || __sync_val_compare_and_swap(slot, current, value) == current) {
-            break;
-        }
-    } while (true);
-}
-
-template<typename NEXT, typename T>
-inline T atomic_next(volatile T *val, NEXT next) {
-    do {
-        T current = __atomic_load_n(val, __ATOMIC_RELAXED);
-        T n = next(current);
-        if (__sync_val_compare_and_swap(val, current, n) == current) {
-            return n;
-        }
-    } while (true);
-}
-
 template<typename T>
 class txn_scoreboard_t {
     uint32_t mask = 0;
     uint32_t size = 0;
-    int64_t max = 0;
-    // 1-based min txn that is in-use
-    // we have to use 1 based txn to rule out possibility of 0 txn
-    // 0 is initial value when shared memory is created
-    int64_t min = L_MAX;
-    T counts[];
-
-    inline static T inc(T val) {
-        return val + 1;
+    std::atomic<int64_t> max{0};
+    std::atomic<int64_t> min{0};
+    std::atomic<T> counts[];
+
+    template<typename TT>
+    inline static TT set_max_atomic(std::atomic<TT> &slot, TT value) {
+        TT current = slot.load();
```

Comment:
nit: we could make use of std::memory_order. You may find a suggested change here:
https://github.com/puzpuzpuz/questdb/commit/6ab0a22b02d39b44322805d75b065cc7f88663ba

Reply:
I'd rather skip memory model optimisation here. It not going to yield much, it's not super hot path + JNI calls add overhead more than memory model relaxation

Reply:
Sure, let's go with sequential consistency.

Reply:
Did quick and dirty benchmarking to verify that the scoreboard operations aren't way too expensive.

The scenario is the following: readers only acquiring and releasing the same transaction number. Single thread - ~40ns, 8 threads - ~400ns. The increased cost under contention is expected and the performance is ok considering that read-only query latencies are beyond ms.